node.cpp

```diff
@@ -215,21 +246,39 @@ bool ThreadedNodeBase::startThread() {
 void ThreadedNodeBase::stop() {
     thread_must_quit_ = true;
     NodeBase::stop();
-    thread_.waitForTermination();
+    if (creating_thread_id_ == std::this_thread::get_id()) {
+      thread_.waitForTermination();
+    }
 }
 
 bool ThreadedNodeBase::isRunning() const {
     return NodeBase::isRunning() && thread_.isRunning();
 }
 
+void ThreadedNodeBase::waitUntilStopped() {
+    NodeBase::waitUntilStopped();
+    if (creating_thread_id_ == std::this_thread::get_id()) {
+      thread_.waitForTermination();
+    }
+}
+
 void ThreadedNodeBase::threadEntryPoint(void *ptr) {
     ThreadedNodeBase* instance = static_cast<ThreadedNodeBase*>(ptr);
 
 #ifdef MEDIAGRAPH_USE_EASY_PROFILER
     EASY_THREAD(instance->name().c_str());
 #endif
 
-    instance->threadMain();
+    try {
+        instance->threadMain();
+    } catch (std::exception& e) {
+        std::cerr << instance->name() << ": exception: "
```

Comment:
it may be clearer to say that this happened in a mediagraph node, as it can be confused with other log messages.

What about
```suggestion
        std::cerr << "Mediagraph thread exception at node <" << instance->name() << ">: "
```

Reply:
changed it to:
```
        std::cerr << "Uncaught top-level exception in mediagraph thread, node: "
            << instance->name() << ": exception: " << e.what() << std::endl;
```

do you think I should call std::terminate() or std::abort() right after this?

Reply:
in the current version, the node finishes, and will that trigger the graph to stop?
If so, then I'm ok with the current behavior.

Otherwise I'd propose to add a way to let the graph know if it needs to be stopped from a node, though it's not urgent.

Reply:
OK so let's try like this, we'll change it later if needed.